### DISCUSSION TITLE
How can we improve NanoChat to make its performance on the HumanEval and GSM8K benchmarks approach that of Qwen?

### DISCUSSION THREAD

**Discussion by devymex** [karpathy/nanochat#167] [Q&A] [answered]
<img width="628" height="199" alt="image" src="https://github.com/user-attachments/assets/1b3f1f89-0686-4d09-8440-e75152786edc" />

\* from https://arxiv.org/pdf/2505.02214v1

**Comment by MichelNivard** [comment]
I guess way more data, especially code data. There is a really great filtered high quality code dataset by the huggingface team: https://huggingface.co/datasets/HuggingFaceTB/stack-edu

Blend this is with the fineweb (making sure to keep a seriously large chunk natural language not just code). Then if that's not enough you could consider writing custom SFT and/or reinforcement learning based on coding. Think Q and A where the Q is a stack exchange like question and the RL reward is based on the awnser containing fenced code and if you want to be really fancy about it testing and scoring wether the fenced code running without error. Maybe the RL isn't needed at all and more specific training data does the job...

**Comment by karpathy** [comment, marked as answer]
So one tricky thing to deal with design-wise of the repo is that I believe these tiny models by Qwen and others are all distillations. They are simply trained to mimic their much larger cousins in the miniseries. Therefore, the way you train the small model is not the way you'd train the largest model. For example, I'm not sure that they do something that looks like pretraining at all. I'm trying to make nanochat resemble algorithmically how you'd train a giant model. However, that means that we're also doing suboptimal things as far as top benchmark performance alone is involved. If we chase benchmark performance (as obviously we do to some extent), I'm afraid we'll slowly converge ourselves to basically a big model distillation pipeline.

**Reply by MichelNivard** [reply]
Hm, but they did release "based" models for the smaller qwen3 models? Would the procedure be to distill the big base into the little base and the big post-trained modle into the small post-trained model? Small models tht are clearly trained ground up like smollm2/smollm3 sort of do almost as well as the smaller qwen3 base models models on base modle liek evals, but after post training qwen does a lot better relatively...

[a comment by an automated account is not shown]
